pgxn/neon/walproposer.c

```diff
@@ -723,13 +728,176 @@ SendProposerGreeting(Safekeeper *sk)
 	BlockingWrite(sk, sk->outbuf.data, sk->outbuf.len, SS_HANDSHAKE_RECV);
 }
 
+/*
+ * Assuming `sk` sent its node id, find such member(s) in wp->mconf and set ptr in
+ * members_safekeepers & new_members_safekeepers to sk.
+ */
+static void
+UpdateMemberSafekeeperPtr(WalProposer *wp, Safekeeper *sk)
+{
+	/* members_safekeepers etc are fixed size, sanity check mconf size */
+	if (wp->mconf.members.len > MAX_SAFEKEEPERS)
+		wp_log(FATAL, "too many members %d in mconf", wp->mconf.members.len);
+	if (wp->mconf.new_members.len > MAX_SAFEKEEPERS)
+		wp_log(FATAL, "too many new_members %d in mconf", wp->mconf.new_members.len);
+
+	/* node id is not known until greeting is received */
+	if (sk->state < SS_WAIT_VOTING)
+		return;
+
+	/* 0 is assumed to be invalid node id, should never happen */
```

Comment:
right now we do support 0 as valid node id in the storage controller, and 0 does get assigned to storcons in pytests. does other code assume 0 to be invalid?

It's not hard to change it in pytests though, just needs to be changed here: https://github.com/neondatabase/neon/pull/11261/files#diff-9ffba544d49919e927a416b899e92d769f37f3c8355258eb212678b4e5ca067bR1331

Reply:
Maybe also check it in the `handle_upsert_safekeeper` function that negative and zero ids can't be registered.

Reply:
Good catch. In principle there is no hard requirement for zero to be invalid, but it makes sense to be consistent in all places. I added the check here because C (and sk <-> wp proto) isn't great at Options and differentiating zero / invalid value.

Since for a long time we issue node ids from 1 in neon_local (and there are no 0s issued by cplane as well), I'd rather forbid zeros everywhere.

Reply:
resolved via https://github.com/neondatabase/neon/pull/11419